graphql_compiler/interpreter/typedefs.py

```diff
@@ -172,8 +172,73 @@ def project_property(
         neighbor_hints: Optional[Collection[Tuple[EdgeInfo, NeighborHint]]] = None,
         **hints: Any,
     ) -> Iterable[Tuple[DataContext[DataToken], Any]]:
-        """Produce the values for a given property for each of an iterable of input DataTokens."""
-        # TODO(predrag): Add more docs in an upcoming PR.
+        """Produce the values for a given property for each of an iterable of input DataTokens.
+
+        In situations such as outputting property values or applying filters to properties,
+        the interpreter needs to get the value of some property field for a series of DataTokens.
+        For example, consider the following query:
+            {
+                Foo {
+                    bar @output(out_name: "bar_value")
+                }
+            }
+        Once the interpreter has used the get_tokens_of_type() function to obtain
+        an iterable of DataTokens for the Foo type, it will need to get the value of
+        the "bar" property on those tokens. This is where project_property() comes in.
+        The interpreter would call project_property() with arguments current_type_name = "Foo"
+        and field_name = "bar" along with an iterable of DataContext values, each of which
+        contains an optional DataToken as its current_token attribute. These DataContext
+        objects simply allow the interpreter to keep track of "which data came from where";
```

Comment:
Consider including this bit about what `DataContexts` are earlier in the doc string. I think the difference between `DataContext` and `DataToken` is not immediately obvious without this and the reader might be confused until they get to this point.

Reply:
Should be [coming soon](https://github.com/kensho-technologies/graphql-compiler/blob/936e47970eff81c2a6eeaead44fdf4d8d3c1a599/graphql_compiler/interpreter/typedefs.py#L23)

Reply:
Upon reading this one more time, I think I agree with @chewselene. While the reader could refer to the (not-yet-written) docs for `DataContext`, it's entirely possible that they miss the fact that `DataToken` and `DataContext` are not the same thing, and get confused. If we can make this part stand better on its own, we should. Let me give it another shot.

Reply:
Tweaked this, let me know what you think!